### ISSUE TITLE
prof2 shows call path within GPU kernels in trace when it shouldn't

### ISSUE THREAD

**Issue by jmellorcrummey** [HPCToolkit/hpctoolkit#619]
When measuring and analyzing LAMMPS with prof2 in the johnmc branch, I found that prof2 is adding detail about call paths internal to the GPU to the heterogeneous call chain for a kernel call in the GPU trace. We don't trace information within GPU kernels so we can't display it. 

The figure below illustrates the problem. Notice that the trace shows a deep call path WITHIN the GPU kernel so that all of the elements in the trace appear to be in `<unknown procedure>` (green at the bottom of the call chain) rather than in `Kokkos::impl::cuda_parallel_launch_constant_memory<...>() `(the item on the path immediately below `<gpu kernel>`. Even when PC sampling, we don't trace internal details of kernels because we get GPU PC samples as a histogram (with the time dimension collapsed out).

<img width="1395" alt="image" src="https://user-images.githubusercontent.com/3926112/177062982-69286233-cd74-419b-8054-1b595a395aa1.png">

The database that yielded this figure can be found in ufront.cs.rice.edu:~johnmc/bugs/prof2-trace-bug.tgz

**Comment by blue42u** [comment]
These parts of the trace refer to the IP `d0cb001a8cc42528f80012cf1143294c.gpubin+0x8f880`, which the Structfile maps to a long inlined call stack ending in `<unknown procedure>`:
```xml
<P i="623" n="Kokkos::Impl::cuda_parallel_launch_constant_memory&lt;Kokkos::Impl::ParallelReduce&lt;LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;, Kokkos::RangePolicy&lt;Kokkos::Cuda, LAMMPS_NS::TagComputeTempScalar&lt;0&gt; &gt;, Kokkos::InvalidType, Kokkos::Cuda&gt; &gt;() [d0cb001a8cc42528f80012cf1143294c.gpubin]" ln="_ZN6Kokkos4Impl36cuda_parallel_launch_constant_memoryINS0_14ParallelReduceIN9LAMMPS_NS17ComputeTempKokkosINS_4CudaEEENS_11RangePolicyIJS5_NS3_20TagComputeTempScalarILi0EEEEEENS_11InvalidTypeES5_EEEEvv" s="17" l="102" v="{[0x8f880-0x8f881)}">
<A i="625" l="106" f=".../Kokkos_Cuda_KernelLaunch.hpp" n="" v="{}">
    <A i="626" l="1041" f=".../Kokkos_Cuda_Parallel.hpp" n="Kokkos::Impl::ParallelReduce&lt;LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;, Kokkos::RangePolicy&lt;Kokkos::Cuda, LAMMPS_NS::TagComputeTempVector&lt;1&gt; &gt;, Kokkos::InvalidType, Kokkos::Cuda&gt;::operator()() const" v="{}">
    <A i="627" l="1041" f=".../Kokkos_Cuda_Parallel.hpp" n="" v="{}">
        <A i="628" l="883" f=".../Kokkos_Cuda_ReduceScan.hpp" n="Kokkos::Impl::cuda_single_inter_block_reduce_scan&lt;false, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;, LAMMPS_NS::TagComputeTempVector&lt;1&gt;, unsigned int&gt;(LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt; const&amp;, unsigned int, unsigned int, unsigned int*, unsigned int*, unsigned int*)" v="{}">
        <A i="629" l="883" f=".../Kokkos_Cuda_ReduceScan.hpp" n="" v="{}">
            <A i="630" l="512" f=".../Kokkos_Cuda_ReduceScan.hpp" n="Kokkos::Impl::CudaReductionsFunctor&lt;LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;, LAMMPS_NS::TagComputeTempVector&lt;1&gt;, false, true&gt;::scalar_inter_block_reduction(LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt; const&amp;, unsigned int, unsigned int, unsigned int*, unsigned int*, unsigned int*)" v="{}">
            <A i="631" l="512" f=".../Kokkos_Cuda_ReduceScan.hpp" n="" v="{}">
                <A i="632" l="472" f=".../Kokkos_Cuda_ReduceScan.hpp" n="Kokkos::Impl::CudaReductionsFunctor&lt;LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;, LAMMPS_NS::TagComputeTempVector&lt;1&gt;, false, true&gt;::scalar_intra_block_reduction(LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt; const&amp;, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP, bool, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP*, int, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP*)" v="{}">
                <A i="633" l="472" f=".../Kokkos_Cuda_ReduceScan.hpp" n="" v="{}">
                    <A i="634" l="437" f=".../Kokkos_Cuda_ReduceScan.hpp" n="Kokkos::Impl::CudaReductionsFunctor&lt;LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;, LAMMPS_NS::TagComputeTempVector&lt;1&gt;, false, true&gt;::scalar_intra_warp_reduction(LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt; const&amp;, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP, bool, int, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP&amp;)" v="{}">
                    <A i="635" l="437" f=".../Kokkos_Cuda_ReduceScan.hpp" n="" v="{}">
                        <A i="636" l="170" f=".../Kokkos_Cuda_Vectorization.hpp" n="Kokkos::Impl::in_place_shfl&lt;LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP&amp;, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP&amp;, int, int const&amp;, unsigned int&amp;&gt;(LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP&amp;, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP&amp;, int&amp;&amp;, int const&amp;, unsigned int&amp;)" v="{}">
                        <A i="637" l="170" f=".../Kokkos_Cuda_Vectorization.hpp" n="" v="{}">
                            <A i="638" l="149" f=".../Kokkos_Cuda_Vectorization.hpp" n="Kokkos::Impl::in_place_shfl_op&lt;Kokkos::Impl::in_place_shfl_fn&gt;::operator()&lt;LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP&gt;(LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP&amp;, LAMMPS_NS::ComputeTempKokkos&lt;Kokkos::Cuda&gt;::s_CTEMP const&amp;, int, int, unsigned int) const" v="{}">
                            <A i="639" l="149" f=".../Kokkos_Cuda_Vectorization.hpp" n="" v="{}">
                                <A i="640" l="165" f=".../Kokkos_Cuda_Vectorization.hpp" n="Kokkos::Impl::in_place_shfl_fn::do_shfl_op&lt;int const&gt;(unsigned int, int const&amp;, int, int) const" v="{}">
                                <A i="641" l="165" f=".../Kokkos_Cuda_Vectorization.hpp" n="" v="{}">
                                    <A i="642" l="373" f=".../sm_30_intrinsics.hpp" n="_INTERNAL_89f6ff90_23_compute_temp_kokkos_cpp_7fd59dcb_2660559::__shfl_sync(unsigned int, int, int, int)" v="{}">
                                    <A i="643" l="373" f=".../sm_30_intrinsics.hpp" n="" v="{}">
                                        <A i="644" l="102" f=".../Kokkos_Cuda_KernelLaunch.hpp" n="&lt;unknown procedure&gt;" v="{}">
                                            <S i="645" l="102" v="{[0x8f880-0x8f890)}" />
```
So the `<unknown procedure>` part is a case of "garbage in, garbage out."

Otherwise, this is very similar to 1d7b1a38a31a55f8d8786e3a368f6dec166d279d (couldn't find the PR for it). Based on that, possible solutions:
 - Add a flag to the trace sample structure to indicate that the sample refers to the GPU kernel (i.e. kernel-launch-edge) instead of the instruction itself. Adding one flag byte will increase the size of a measurement trace by 8%, two by 16%, etc.
 - Add a flag to the trace header to indicate that all samples in the trace refer to the GPU kernel instead of the instructions themselves. This may require some care in Run to ensure edge-traces are not mixed with instruction-traces.
 - Add logic in Prof2 to consider trace-points just below a `<gpu kernel>` placeholder as referring to the GPU kernel instead of the instruction itself. This will make adding "real" GPU traces later much harder (and add unpreferred logic "magic" to Prof2).

**Comment by jmellorcrummey** [comment]
I think the solution is the last:

> Add logic in Prof2 to consider trace-points just below a <gpu kernel> placeholder as referring to the GPU kernel instead of the instruction itself. This will make adding "real" GPU traces later much harder (and add unpreferred logic "magic" to Prof2).

We will NEVER have traces at a finer granularity than instructions from GPUs.  At best, we will receive a sequence of snapshots, where a snapshot includes each active or "on deck" PC across the entire GPU. Such a snapshot would indicate where every wave/warp/workgroup-item-set is at a point in time across all the compute units/execution units/streaming multiprocessors. That is basically a histogram at an instant of time. We can't build a finer-grain trace out of those. 

The only way we could get something finer grained is with instrumentation. I don't think it is worth the cost.

**Comment by jmellorcrummey** [comment]
After looking at this a bit more, another alternative is to just use a filter to make `<unknown procedure>` disappear from the bottom of that call chain in the kernel. The rest of the chain indicates where the first instruction came from.

However, the first instruction isn't the only instruction in the kernel. I think focusing on the call chain for the first instruction gives an illusion of precision when there is none.

**Comment by blue42u** [comment]
> After looking at this a bit more, another alternative is to just use a filter to make <unknown procedure> disappear from the bottom of that call chain in the kernel. The rest of the chain indicates where the first instruction came from.

The `<unknown procedure>` is a bug in either Struct or NVIDIA's DWARF, I think the innermost inlined call should probably be a single-alien (i.e. roughly a macro expansion) instead of a double-alien. So I wouldn't go making changes to the Viewer when the bug is elsewhere (we've done that before, but I'd like to keep it limited to *necessary* evils).

> However, the first instruction isn't the only instruction in the kernel. I think focusing on the call chain for the first instruction gives an illusion of precision when there is none.

Completely agreed.

> We will NEVER have traces at a finer granularity than instructions from GPUs. At best, we will receive a sequence of snapshots, where a snapshot includes each active or "on deck" PC across the entire GPU. Such a snapshot would indicate where every wave/warp/workgroup-item-set is at a point in time across all the compute units/execution units/streaming multiprocessors. That is basically a histogram at an instant of time. We can't build a finer-grain trace out of those.

Couldn't we in theory build an instruction trace per workgroup/SM/CU/... using those snapshots? It might not be worth it now, but given the amount of interest in hardware-level performance analysis IMHO we shouldn't discount the possibility this early in the game. (Also I want to avoid adding magic logic in Prof2 if possible, it's a bear to fix when assumptions shift and the magic flops.)

So IMHO, the second option seems the most appropriate, it has minimal cost now and won't cause headaches later.

**Comment by jmellorcrummey** [comment]
> So IMHO, the second option seems the most appropriate, it has minimal cost now and won't cause headaches later.

We can certainly add a flag to the trace header.

**Comment by Jokeren** [comment]
> Couldn't we in theory build an instruction trace per workgroup/SM/CU/... using those snapshots? It might not be worth it now, but given the amount of interest in hardware-level performance analysis IMHO we shouldn't discount the possibility this early in the game. (Also I want to avoid adding magic logic in Prof2 if possible, it's a bear to fix when assumptions shift and the magic flops.)

That could be very useful for large kernels.

However, it is super expensive, which could cause hundreds of overhead with gigabytes of measurement data.

**Comment by blue42u** [comment]
> However, it is super expensive, which could cause hundreds of overhead with gigabytes of measurement data.

I would think this would generate roughly the same data volume as ~8-16 CPU threads, assuming the snapshots happen on a regular interval (e.g. 500us). That's well within what Prof2 (should) handle.

**Comment by jmellorcrummey** [comment]
resolved by #622

**Comment by Jokeren** [comment]
@blue42u I think prof2 can handle thousands of threads without any problem. On the CU/SM level, it would be equivalent of >100 CPU threads per GPU, considering A100 with 108 SMs; on the scheduler level, it could be > 400 CPU threads. However, I was worried about the runtime (hpcrun) overhead but not postmortem analysis overhead, taking a snapshot or copying large amount of data from the GPU to the CPU is expensive.
